File: mcp_server_devnot/mcp_server_dotnet/Services/ProjectService.cs
using System.Diagnostics;
using System.Text.Json;
using PmMcpServer.Data;

namespace PmMcpServer.Services;

public class ProjectService
{
    private readonly ProjectData _data;
    private static readonly string LogFile = Path.Combine(
        Directory.GetCurrentDirectory(), "benchmark_results.txt");

    public ProjectService()
    {
        var path = Path.Combine(
            AppContext.BaseDirectory, "Data", "project_data.json");
    
        if (!File.Exists(path))
        {
            // fallback: proje dizini
            path = Path.Combine(
                Directory.GetCurrentDirectory(), "Data", "project_data.json");
        }
    
        var json = File.ReadAllText(path);
        var options = new JsonSerializerOptions
        {
            PropertyNameCaseInsensitive = true,
            PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower
        };
        _data = JsonSerializer.Deserialize<ProjectData>(json, options)!;
    }

    private void Log(string toolName, double executionMs)
    {
        var separator = new string('=', 60);
        File.AppendAllText(LogFile,
            $"\n{separator}\n" +
            $"[{DateTime.Now:HH:mm:ss}] {toolName}\n" +
            $"  Execution : {executionMs} ms\n");
    }

    public object GetSprintStatus(string sprintId = "active")
    {
        var sw = Stopwatch.StartNew();
        Sprint? sprint = sprintId == "active"
            ? _data.Sprints.FirstOrDefault(s => s.Status == "active")
            : _data.Sprints.FirstOrDefault(s =>
                s.Id.Equals(sprintId, StringComparison.OrdinalIgnoreCase) ||
                s.Name.Equals(sprintId, StringComparison.OrdinalIgnoreCase));

        sw.Stop();
        Log("GetSprintStatus", Math.Round(sw.Elapsed.TotalMilliseconds, 2));

        if (sprint is null)
            return new { success = false, message = $"Sprint not found: {sprintId}" };

        var done = sprint.Tasks.Count(t => t.Status == "done");
        var inProgress = sprint.Tasks.Count(t => t.Status == "in_progress");
        var todo = sprint.Tasks.Count(t => t.Status == "todo");

        return new
        {
            success = true,
            execution_ms = Math.Round(sw.Elapsed.TotalMilliseconds, 2),
            sprint = new
            {
                sprint.Id, sprint.Name, sprint.Status,
                sprint.Goal, sprint.StartDate, sprint.EndDate,
                summary = new { total_tasks = sprint.Tasks.Count, done, in_progress = inProgress, todo },
                tasks = sprint.Tasks
            }
        };
    }

    public object GetTeamVelocity(int lastNSprints = 4)
    {
        var sw = Stopwatch.StartNew();
        var history = _data.VelocityHistory
            .Where(v => v.Completed.HasValue)
            .TakeLast(lastNSprints).ToList();

        var avg = history.Count > 0
            ? Math.Round(history.Average(v => v.Completed!.Value), 1) : 0;

        string trend = "stable";
        if (history.Count >= 2)
        {
            var last = history[^1].Completed!.Value;
            var prev = history[^2].Completed!.Value;
            trend = last > prev ? "improving" : last < prev ? "declining" : "stable";
        }

        sw.Stop();
        Log("GetTeamVelocity", Math.Round(sw.Elapsed.TotalMilliseconds, 2));

        return new
        {
            success = true,
            execution_ms = Math.Round(sw.Elapsed.TotalMilliseconds, 2),
            velocity = new { average = avg, history, trend }
        };
    }

    public object GetBacklogItems(string priority = "all", int limit = 5)
    {
        var sw = Stopwatch.StartNew();
        var items = _data.Backlog.AsEnumerable();

        if (priority != "all")
            items = items.Where(i => i.Priority == priority.ToLower());

        var order = new Dictionary<string, int> { ["high"] = 0, ["medium"] = 1, ["low"] = 2 };
        var result = items.OrderBy(i => order.GetValueOrDefault(i.Priority, 99))
                          .Take(limit).ToList();

        sw.Stop();
        Log("GetBacklogItems", Math.Round(sw.Elapsed.TotalMilliseconds, 2));

        return new
        {
            success = true,
            execution_ms = Math.Round(sw.Elapsed.TotalMilliseconds, 2),
            backlog = new
            {
                count = result.Count,
                filter = priority,
                items = result,
                total_points = result.Sum(i => i.StoryPoints)
            }
        };
    }

    public object GetTeamMembers(string role = "all")
    {
        var sw = Stopwatch.StartNew();
        var members = _data.Team.AsEnumerable();

        if (role != "all")
            members = members.Where(m =>
                m.Role.Contains(role, StringComparison.OrdinalIgnoreCase));

        var result = members.ToList();
        sw.Stop();
        Log("GetTeamMembers", Math.Round(sw.Elapsed.TotalMilliseconds, 2));

        return new
        {
            success = true,
            execution_ms = Math.Round(sw.Elapsed.TotalMilliseconds, 2),
            team = new { count = result.Count, members = result }
        };
    }

    public object GetTasksByAssignee(string assignee)
    {
        var sw = Stopwatch.StartNew();
        var tasks = _data.Sprints
            .SelectMany(s => s.Tasks
                .Where(t => t.Assignee.Equals(assignee, StringComparison.OrdinalIgnoreCase))
                .Select(t => new { t.Id, t.Title, t.Assignee, t.Status, t.StoryPoints, Sprint = s.Name }))
            .ToList();

        sw.Stop();
        Log("GetTasksByAssignee", Math.Round(sw.Elapsed.TotalMilliseconds, 2));

        if (tasks.Count == 0)
            return new { success = false, message = $"No tasks found for: {assignee}" };

        return new
        {
            success = true,
            execution_ms = Math.Round(sw.Elapsed.TotalMilliseconds, 2),
            assignee, count = tasks.Count, tasks
        };
    }
}

File: mcp_python/services/project_service.py
import json
import time
from pathlib import Path
from typing import Any
from datetime import datetime

LOG_FILE = Path(__file__).parent.parent / "benchmark_results.txt"

_data = None

def _load():
    global _data
    if _data is None:
        path = Path(__file__).parent.parent / "data" / "project_data.json"
        _data = json.loads(path.read_text(encoding="utf-8"))
    return _data

def _log(tool_name: str, execution_ms: float):
    with open(LOG_FILE, "a", encoding="utf-8") as f:
        f.write(f"\n{'='*60}\n")
        f.write(f"[{datetime.now().strftime('%H:%M:%S')}] {tool_name}\n")
        f.write(f"  Execution : {execution_ms} ms\n")

def get_sprint_status(sprint_id: str = "active") -> dict[str, Any]:
    t0 = time.perf_counter()
    data = _load()

    if sprint_id == "active":
        sprint = next((s for s in data["sprints"] if s["status"] == "active"), None)
    else:
        sprint = next(
            (s for s in data["sprints"]
             if s["id"].lower() == sprint_id.lower() 
             or s["name"].lower() == sprint_id.lower()),
            None
        )

    if not sprint:
        return {"success": False, "message": f"Sprint not found: {sprint_id}"}

    done = [t for t in sprint["tasks"] if t["status"] == "done"]
    in_progress = [t for t in sprint["tasks"] if t["status"] == "in_progress"]
    todo = [t for t in sprint["tasks"] if t["status"] == "todo"]

    execution_ms = round((time.perf_counter() - t0) * 1000, 2)
    _log("get_sprint_status", execution_ms)

    return {
        "success": True,
        "execution_ms": execution_ms,
        "sprint": {
            "id": sprint["id"],
            "name": sprint["name"],
            "status": sprint["status"],
            "goal": sprint["goal"],
            "start_date": sprint["start_date"],
            "end_date": sprint["end_date"],
            "summary": {
                "total_tasks": len(sprint["tasks"]),
                "done": len(done),
                "in_progress": len(in_progress),
                "todo": len(todo)
            },
            "tasks": sprint["tasks"]
        }
    }

def get_team_velocity(last_n_sprints: int = 4) -> dict[str, Any]:
    t0 = time.perf_counter()
    data = _load()

    history = [v for v in data["velocity_history"] if v["completed"] is not None]
    history = history[-last_n_sprints:]
    avg = round(sum(v["completed"] for v in history) / len(history), 1) if history else 0

    trend = "stable"
    if len(history) >= 2:
        last = history[-1]["completed"]
        prev = history[-2]["completed"]
        trend = "improving" if last > prev else "declining" if last < prev else "stable"

    execution_ms = round((time.perf_counter() - t0) * 1000, 2)
    _log("get_team_velocity", execution_ms)

    return {
        "success": True,
        "execution_ms": execution_ms,
        "velocity": {"average": avg, "history": history, "trend": trend}
    }

def get_backlog_items(priority: str = "all", limit: int = 5) -> dict[str, Any]:
    t0 = time.perf_counter()
    data = _load()

    items = data["backlog"]
    if priority != "all":
        items = [i for i in items if i["priority"] == priority.lower()]

    items = sorted(items, key=lambda x: {"high": 0, "medium": 1, "low": 2}[x["priority"]])
    items = items[:limit]

    execution_ms = round((time.perf_counter() - t0) * 1000, 2)
    _log("get_backlog_items", execution_ms)

    return {
        "success": True,
        "execution_ms": execution_ms,
        "backlog": {
            "count": len(items),
            "filter": priority,
            "items": items,
            "total_points": sum(i["story_points"] for i in items)
        }
    }

def get_team_members(role: str = "all") -> dict[str, Any]:
    t0 = time.perf_counter()
    data = _load()

    members = data["team"]
    if role != "all":
        members = [m for m in members if role.lower() in m["role"].lower()]

    execution_ms = round((time.perf_counter() - t0) * 1000, 2)
    _log("get_team_members", execution_ms)

    return {
        "success": True,
        "execution_ms": execution_ms,
        "team": {"count": len(members), "members": members}
    }

def get_tasks_by_assignee_sprint(assignee: str) -> dict[str, Any]:
    t0 = time.perf_counter()
    data = _load()

    tasks = []
    for sprint in data["sprints"]:
        for task in sprint["tasks"]:
            if task["assignee"].lower() == assignee.lower():
                tasks.append({**task, "sprint": sprint["name"]})

    execution_ms = round((time.perf_counter() - t0) * 1000, 2)
    _log("get_tasks_by_assignee_sprint", execution_ms)

    if not tasks:
        return {"success": False, "message": f"No tasks found for: {assignee}"}

    return {
        "success": True,
        "execution_ms": execution_ms,
        "assignee": assignee,
        "count": len(tasks),
        "tasks": tasks
    }